pinot-core/src/main/java/org/apache/pinot/core/operator/query/SelectionPartiallyOrderedByDescOperation.java

```diff
@@ -0,0 +1,126 @@
+ */
+public class SelectionPartiallyOrderedByDescOperation extends LinearSelectionOrderByOperator {
+
+  private static final String EXPLAIN_NAME = "SELECT_PARTIAL_ORDER_BY_DESC";
+
+  private int _numDocsScanned = 0;
+  private long _numEntriesScannedPostFilter = 0;
+
+  public SelectionPartiallyOrderedByDescOperation(IndexSegment indexSegment, QueryContext queryContext,
+      List<ExpressionContext> expressions, TransformOperator transformOperator, int sortedExpr) {
+    super(indexSegment, queryContext, expressions, transformOperator, sortedExpr);
+    assert queryContext.getOrderByExpressions() != null;
+    Preconditions.checkArgument(queryContext.getOrderByExpressions().stream()
+            .filter(expr -> expr.getExpression().getType() == ExpressionContext.Type.IDENTIFIER)
+            .findFirst()
+            .orElseThrow(() -> new IllegalArgumentException("The query is not order by identifiers"))
+            .isDesc(),
+        "%s can only be used when the first column in order by is DESC", EXPLAIN_NAME);
+  }
+
+  @Override
+  protected List<Object[]> fetch(Supplier<ListBuilder> listBuilderSupplier) {
+
+    // Ideally we would use a descending cursor, but we don't actually have them
+    // Alternatively, we could store all blocks in a list and iterate them in reverse order, but ProjectionBlocks share
+    // the same DataBlockCache, so they may be ephemeral and being overridden by the next block.
+    // The only alternative we have right now is to retrieve the last LIMIT elements from each block
+
+    TransformBlock block;
+    int numColumnsProjected = _transformOperator.getNumColumnsProjected();
+    int numExpressions = _expressions.size();
+    BlockValSet[] blockValSets = new BlockValSet[numExpressions];
+
+    List<Object[]> localBestRows = new ArrayList<>();
+
+    try {
+      while ((block = _transformOperator.nextBlock()) != null) {
+        int numDocsFetched = block.getNumDocs();
+        _numDocsScanned += numDocsFetched;
+
+        IntFunction<Object[]> rowFetcher = fetchBlock(block, blockValSets);
+
+        ListBuilder listBuilder = listBuilderSupplier.get();
+
+        if (block.getNumDocs() == 0) {
+          continue;
+        }
+
+        // first, calculate the best rows on this block
+        boolean newPartition = false;
+        for (int docId = block.getNumDocs() - 1; docId >= 0; docId--) {
+          newPartition = listBuilder.add(rowFetcher.apply(docId));
+          if (newPartition && listBuilder.sortedSize() >= _numRowsToKeep) {
+            // We changed to a new partition and we have more values than required.
+            // Therefore, we can stop the execution of this block here.
+            break;
+          }
+        }
+        // then try to add the best rows from previous block
+        Iterator<Object[]> localBestRowIt = localBestRows.iterator();
+        while (localBestRowIt.hasNext() && !newPartition && listBuilder.sortedSize() < _numRowsToKeep) {
```

Comment:
This logic doesn't seem correct to me. E.g. when it is new partition, but not enough rows collected, we should still add the row.